internal/datastore/mysql/datastore.go

```diff
@@ -0,0 +1,648 @@
+	query, args, err := sb.Delete(tableName).Where(filter).Limit(batchDeleteSize).ToSql()
+	if err != nil {
+		return -1, err
+	}
+
+	var deletedCount int64
+	for {
+		cr, err := mds.db.ExecContext(ctx, query, args...)
+		if err != nil {
+			return deletedCount, err
+		}
+
+		rowsDeleted, err := cr.RowsAffected()
+		if err != nil {
+			return deletedCount, err
+		}
+		deletedCount += rowsDeleted
+		if rowsDeleted < batchDeleteSize {
+			break
+		}
+	}
+
+	return deletedCount, nil
+}
+
+// IsReady returns whether the datastore is ready to accept data. Datastores that require
+// database schema creation will return false until the migrations have been run to create
+// the necessary tables.
+//
+// fundamentally different from PSQL implementation:
+// - checking if the current migration version is compatible is implemented with IsHeadCompatible
+// - Database seeding is handled here, so that we can decouple schema migration from data migration
+//   and support skeema-based migrations.
+func (mds *Datastore) IsReady(ctx context.Context) (bool, error) {
+	if err := mds.db.PingContext(ctx); err != nil {
+		return false, err
+	}
+
+	currentMigrationRevision, err := mds.driver.Version()
+	if err != nil {
+		return false, err
+	}
+
+	compatible, err := migrations.Manager.IsHeadCompatible(currentMigrationRevision)
+	if err != nil {
+		return false, err
+	}
+	if !compatible {
+		return false, nil
+	}
+
+	// seed base transaction if not present
+	txRevision, err := mds.HeadRevision(ctx)
+	if err != nil {
+		return false, err
+	}
+	if txRevision == datastore.NoRevision {
+		baseRevision, err := mds.seedBaseTransaction(ctx)
+		if err != nil {
+			return false, err
+		}
+		if baseRevision != datastore.NoRevision {
+			// no txRevision here indicates that the base transaction was already seeded when this write
+			// was committed.  only log for the process which actually seeds the transaction.
+			log.Info().Uint64("txRevision", baseRevision.BigInt().Uint64()).Msg("seeded base datastore txRevision")
+		}
+	}
+	return true, nil
+}
+
+// seedBaseTransaction initializes the first transaction revision.
+func (mds *Datastore) seedBaseTransaction(ctx context.Context) (datastore.Revision, error) {
+	ctx, span := tracer.Start(ctx, "seedBaseTransaction")
+	defer span.End()
+
+	tx, err := mds.db.BeginTx(ctx, nil)
+	if err != nil {
+		return datastore.NoRevision, err
+	}
+	defer migrations.LogOnError(ctx, tx.Rollback)
+
+	// idempotent INSERT IGNORE transaction id=1. safe to be execute concurrently.
+	result, err := tx.ExecContext(ctx, mds.createBaseTxn)
+	if err != nil {
+		return datastore.NoRevision, fmt.Errorf("seedBaseTransaction: %w", err)
+	}
+
+	lastInsertID, err := result.LastInsertId()
+	if err != nil {
+		return datastore.NoRevision, fmt.Errorf("seedBaseTransaction: failed to get last inserted id: %w", err)
+	}
+
+	err = tx.Commit()
+	if err != nil {
+		return datastore.NoRevision, err
+	}
+
+	if lastInsertID == 0 {
```

Comment:
move `0` into a constant

Reply:
addressed in https://github.com/authzed/spicedb/pull/525/commits/28e18d9140c5dbbf25f384154174ba5375d05c0f